lib/msal-node/src/client/ClientCredentialClient.ts

```diff
@@ -214,23 +222,41 @@ export class ClientCredentialClient extends BaseClient {
                 thumbprint
             );
 
-            // check if a new token is being requested because the cached token needs to be refreshed, not because it's expired
-            if (accessTokenCacheKey) {
-                // check if ESTS is down (status code 429 or 5xx)
-                if ((response.status === 429) || (response.status >= 500)) {
+            serverTokenResponse = response.body;
+
+            // if a new token is being requested due to the refresh_in value
+            if (accessTokenRefresh) {
+                // if AAD is down
+                if (response.status >= 500) {
+                    // log a message to the user and do nothing - the cached access token is not able to be refreshed
                     this.logger.warning(
-                        "ClientCredentialClient:executeTokenRequest - AAD is currently unavailable and is unable to refresh the token."
+                        "ClientCredentialClient:executeTokenRequest - AAD is currently unavailable and the access token is unable to be refreshed."
                     );
-                // ESTS is not down, the cached token is bad and should be removed from the cache
-                } else {
+
+                    return null;
+                }
+
+                // AAD is not down, the cached token is bad and should be removed from the cache, and an exception should be thrown
+                if ((response.status >= 400) && (response.status < 499)) {
                     this.logger.error(
-                        "ClientCredentialClient:executeTokenRequest - AAD is currently available but is unable to refresh the token. Returning the unexpired token and removing it from the cache."
+                        "ClientCredentialClient:executeTokenRequest - AAD is currently available but is unable to refresh the access token. The access token is bad and must be removed from the cache."
                     );
+
+                    // we already know there is a valid cached access token
+                    const cachedAccessToken = this.readAccessTokenFromCache();
+                    // get its cache key
+                    const accessTokenCacheKey = (cachedAccessToken as AccessTokenEntity).generateCredentialKey();
+                    // and remove it from the cache
                     this.cacheManager.removeAccessToken(accessTokenCacheKey);
+
+                    const errString = `${serverTokenResponse.error_codes} - [${serverTokenResponse.timestamp}]: ${serverTokenResponse.error_description} - Correlation ID: ${serverTokenResponse.correlation_id} - Trace ID: ${serverTokenResponse.trace_id}`;
```

Comment:
This is not the right place to transform server token response into errors. It makes the code brittle as we not have multiple places that do that.